DTO docs (#1569)
litestar-org/litestar: docs/usage/dto/1-dto-factory.rst

DTO Factory
===========

Litestar maintains a suite of DTO factory types that can be used to create DTOs for use with popular data modelling

> libraries, such as ORMs. These take a model type as a generic type argument, and create subtypes of

Reviewer: ```suggestion
libraries, such as SQLAlchemy or Pydantic. These take a model type as a generic type argument, and create subtypes of
```

"ORM" isn't really a popular library so this should either be concrete examples of libraries or simply refer to "data modelling libraries, such as ORMs".

:class:`AbstractDTOFactory <litestar.dto.factory.abc.AbstractDTOFactory>` that support conversion of that model type to
and from raw bytes.

The following factories are currently available:

- :class:`DataclassDTO <litestar.dto.factory.stdlib.DataclassDTO>`
- :class:`SQLAlchemyDTO <litestar.contrib.sqlalchemy.dto.SQLAlchemyDTO>`

Using DTO Factories
-------------------

DTO factories are used to create DTOs for use with a particular data modelling library. The following example creates

> a DTO for use with a SQLAlchemy model:

Reviewer: > **a** SQLAlchemy model:

Now I know how you pronounce SQL :grimacing:

Author: My my, this is controversial - "Sequel-Alchemy" or "Ess-Que-Ell-Alchemy".. I don't know if the community is ready to have this discussion.

.. literalinclude:: /examples/data_transfer_objects/factory/simple_dto_factory_example.py
    :caption: A SQLAlchemy model DTO
    :language: python

Here we see that a SQLAlchemy model is used as both the ``data`` and return annotation for the handler, and while
Litestar does not natively support encoding/decoding to/from SQLAlchemy models, through
:class:`SQLAlchemyDTO <litestar.contrib.sqlalchemy.dto.SQLAlchemyDTO>` we can do this.

However, we do have some issues with the above example. Firstly, the user's password has been returned to them in the
response from the handler. Secondly, the user is able to set the ``created_at`` field on the model, which should only
ever be set once, and defined internally.

> Lets explore how we can configure DTOs to manage scenarios like these.

Reviewer: ```suggestion
Let's explore how we can configure DTOs to manage scenarios like these.
```

Marking fields
--------------

The :func:`dto_field <litestar.dto.factory.dto_field>` function can be used to mark model attributes with DTO-based
configuration.

Fields marked as `"private"` or `"read-only"` will not be parsed from client data into the user model, and `"private"`
fields are never serialized into return data.

.. literalinclude:: /examples/data_transfer_objects/factory/marking_fields.py
    :caption: Marking fields
    :language: python
    :emphasize-lines: 6,14,15
    :linenos:

> Note:

Reviewer: ```suggestion
.. note::
```
